nemo/collections/llm/evaluation/base.py

```diff
@@ -0,0 +1,209 @@
+        if return_text:
+            # in case of generate_until tasks return just the text
+            return response_data['choices'][0]['text']
+
+        if return_logits:
+            # in case of loglikelihood tasks return the logits
+            return response_data['choices'][0]['generation_logits']
+
+    def tokenizer_type(self, tokenizer):
+        """
+        Returns the type of the tokenizer.
+        """
+        if isinstance(tokenizer, AutoTokenizer):
+            return "AutoTokenizer"
+        elif isinstance(tokenizer, SentencePieceTokenizer):
+            return "SentencePieceTokenizer"
+        else:
+            raise ValueError(
+                "Tokenizer type is not one of SentencePieceTokenizer or HF's AutoTokenizer. Please check "
+                "how to handle special tokens for this tokenizer"
+            )
+
+    def loglikelihood(self, requests: list[Instance]):
+        """
+        Defines the loglikelihood request. Takes input requests of type list[Instance] where Instance is a dataclass
+        defined in lm_eval.api.instance. Each Instance conists of the input prompt, output prompt, request type(here
+        loglikelihood) and other relevant args like few shot samples.
+        """
+        special_tokens_kwargs = {}
+        if self.tokenizer_type(self.tokenizer) == "SentencePieceTokenizer":
+            special_tokens_kwargs['add_bos'] = self.add_bos
+        elif self.tokenizer_type(self.tokenizer) == "AutoTokenizer":
+            special_tokens_kwargs['add_special_tokens'] = self.add_bos
+
+        results = []
+        for request in requests:
+            # get the input prompt from the request
+            context = request.arguments[0]
+            # get the output prompt from the request
+            continuation = request.arguments[1]
+            # get encoded tokens of continuation
+            continuation_enc = self.tokenizer.tokenizer.encode(continuation, **special_tokens_kwargs)
+            # for SentencePeice consider the encoded tokens from the 2nd token since first encoded token is space.
+            if self.tokenizer_type(self.tokenizer) == "SentencePieceTokenizer":
+                continuation_enc = continuation_enc[1:]
+            num_cont_tokens = len(continuation_enc)
+            # Update self.max_tokens_to_generate with number of continuation tokens (or output tokens) in the request
+            self.max_tokens_to_generate = num_cont_tokens
+            # Create payload to query the model deployed on PyTriton server
+            payload = {
+                "model": self.model_name,
+                "prompt": context,
+                "max_tokens": self.max_tokens_to_generate,
+                "temperature": self.temperature,
+                "top_p": self.top_p,
+                "top_k": self.top_k,
+            }
+            # Get the logits from the model
+            generation_logits = self._generate_tokens_logits(payload, return_logits=True)
```

Comment:
n00b question: do you need to specify the torch device somewhere for these tensors?

Reply:
The generation_logits we get at this point are already on the CPU. The logits returned from the trtllm model are on cuda device and are moved to cpu [here](https://github.com/NVIDIA/NeMo/pull/10621/files/355243481b4a5e1346358fc8eb57c6d4db1a0c88#diff-3a3bd49eed8a038989253f6f2e0a01cf494d1de391863d4ab2377a1f7d4cf832R920)